src/MixedQuoter.sol

```diff
@@ -0,0 +1,327 @@
+                (int128[] memory deltaAmounts,) = binQuoter.quoteExactInputSingle(
+                    IBinQuoter.QuoteExactSingleParams({
+                        poolKey: binParams.poolKey,
+                        zeroForOne: zeroForOne,
+                        exactAmount: amountIn.toUint128(),
+                        hookData: binParams.hookData
+                    })
+                );
+                amountIn = deltaAmounts[zeroForOne ? 1 : 0].toUint256();
+            } else if (action == MixedQuoterActions.SS_2_EXACT_INPUT_SINGLE) {
+                (tokenIn, tokenOut) = convertNativeToWETH(tokenIn, tokenOut);
+                // params[actionIndex] is zero bytes
+                amountIn = quoteExactInputSingleStable(
+                    QuoteExactInputSingleStableParams({
+                        tokenIn: tokenIn,
+                        tokenOut: tokenOut,
+                        amountIn: amountIn,
+                        flag: 2
+                    })
+                );
+            } else if (action == MixedQuoterActions.SS_3_EXACT_INPUT_SINGLE) {
+                (tokenIn, tokenOut) = convertNativeToWETH(tokenIn, tokenOut);
+                // params[actionIndex] is zero bytes
+                amountIn = quoteExactInputSingleStable(
+                    QuoteExactInputSingleStableParams({
+                        tokenIn: tokenIn,
+                        tokenOut: tokenOut,
+                        amountIn: amountIn,
+                        flag: 3
+                    })
+                );
+            } else {
+                revert UnsupportedAction(action);
+            }
+        }
+
+        return amountIn;
+    }
+
+    function checkV4PoolKeyCurrency(PoolKey memory poolKey, bool isZeroForOne, address tokenIn, address tokenOut)
+        private
+        pure
+    {
+        Currency currency0;
+        Currency currency1;
+        if (isZeroForOne) {
+            currency0 = Currency.wrap(tokenIn);
+            currency1 = Currency.wrap(tokenOut);
+        } else {
+            currency0 = Currency.wrap(tokenOut);
+            currency1 = Currency.wrap(tokenIn);
+        }
+        if (!equals(poolKey.currency0, currency0) || !equals(poolKey.currency1, currency1)) {
+            revert InvalidPoolKeyCurrency();
+        }
+    }
+
+    function convertWETHToV4NativeCurency(bool isWETHPool, PoolKey memory poolKey, address tokenIn, address tokenOut)
+        private
+        view
+        returns (address, address)
+    {
+        if (!isWETHPool && poolKey.currency0.isNative()) {
+            if (tokenIn == WETH9) {
+                tokenIn = ZERO_ADDRESS;
+            }
+            if (tokenOut == WETH9) {
+                tokenOut = ZERO_ADDRESS;
+            }
+        }
+        return (tokenIn, tokenOut);
+    }
+
+    function convertNativeToWETH(address tokenIn, address tokenOut) private view returns (address, address) {
+        if (tokenIn == ZERO_ADDRESS) {
```

Comment:
1/ would it be good to like import `CurrencyLibrary` and do `tokenIn == CurrencyLibrary.Native` instead of `ZERO_ADDRESS`

2/ add in some context on why zero_address was used in as WETH. eg.

```
/// @dev v4 supports native pool and address(0) represent the native token. When quotes across paths (v3 -> v4), tokens might be [usdc, eth, cake] to get a quote for v3 usdc -> weth and v4 eth -> cake. In this case, the eth is address(0) and convert to weth
```